### PR TITLE
Use NodeType enum instead of Strings

### PR DESCRIPTION
**Pull request by leventov**
For type safety and code readability.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/discovery/NodeType.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package org.apache.druid.discovery;
+
+/**
+ * Constant names are lowercase for backward compatibility, as a part of the {@link DiscoveryDruidNode}'s JSON format.
+ */
+public enum NodeType
+{
+  coordinator,
```
Comment (QiuMM) [inline server/src/main/java/org/apache/druid/discovery/NodeType.java]: I think all these enum values in uppercase would be better.

Comment (gianm) [reply server/src/main/java/org/apache/druid/discovery/NodeType.java]: You can customize this with a `@JsonValue` annotation: see LongEncodingStrategy for an example.

Comment (leventov) [reply server/src/main/java/org/apache/druid/discovery/NodeType.java]: Right, changed.

File: extensions-core/druid-basic-security/src/main/java/org/apache/druid/security/basic/CommonCacheNotifier.java [hunk 2]
```diff
@@ -57,14 +58,7 @@
...
 
-  private static final List<String> NODE_TYPES = Arrays.asList(
-      DruidNodeDiscoveryProvider.NODE_TYPE_BROKER,
-      DruidNodeDiscoveryProvider.NODE_TYPE_OVERLORD,
-      DruidNodeDiscoveryProvider.NODE_TYPE_HISTORICAL,
-      DruidNodeDiscoveryProvider.NODE_TYPE_PEON,
-      DruidNodeDiscoveryProvider.NODE_TYPE_ROUTER,
-      DruidNodeDiscoveryProvider.NODE_TYPE_MM
-  );
+  private static final List<NodeType> NODE_TYPES = Arrays.asList(NodeType.values());
```
Comment (leventov) [inline extensions-core/druid-basic-security/src/main/java/org/apache/druid/security/basic/CommonCacheNotifier.java]: BTW not sure about this, if coordinator was skipped on purpose or by mistake. See https://github.com/apache/incubator-druid/pull/5099#discussion_r220149658. @jon-wei

**Comment by jihoonson** [comment]
There is something similar to this called `ServerType`. Can we merge that with `NodeType`?

File: extensions-core/druid-basic-security/src/main/java/org/apache/druid/security/basic/CommonCacheNotifier.java (hunk 2 above)
Comment (leventov) [reply extensions-core/druid-basic-security/src/main/java/org/apache/druid/security/basic/CommonCacheNotifier.java]: Reverted according to https://github.com/apache/incubator-druid/pull/5099#discussion_r220330380

**Comment by leventov** [comment]
@jihoonson see #6384, I think this is not reasonable unless we are willing to break JSON APIs.

**Comment by leventov** [comment]
@himanshug could you please take a look?

**Comment by himanshug** [comment]
This looks like a trivial change. But, have you tested this on a running cluster to ensure it doesn't break anything?

LGTM otherwise.

**Comment by leventov** [comment]
@himanshug we never test any changes before merging into master. Integration tests are supposed to test inter-node and ZK communications.

**Comment by himanshug** [comment]
I wish I was as confident about integration tests. In general, I think I try and test everything at least in a toy cluster when sending the PR and saying it was done from my side. Even that doesn't catch all problems but gives me as PR author some confidence.

**Comment by himanshug** [comment]
that said, this PR looks pretty trivial and it is likely that integration env will cover it but it is usually nice to test such changes.

**Comment by leventov** [comment]
> I think I try and test everything at least in a toy cluster when sending the PR and saying it was done from my side.

I think this is inefficient. A PR like this one is normally done in less than one hour, while creating a distribution, a test cluster, and testing the new code reasonably is much more work to do. The release baking process with release candidates should provide such testing "in batch" for all PRs that are included in a specific release.

**Comment by himanshug** [comment]
my two cents...
I certainly don't expect all PRs to go through real cluster testing except the ones that change cluster coordination or other things that are hard to unit test etc.
Release baking process is great but doesn't have to be a substitute for testing our own code. Also, by that time a lot of PRs pile on top of each other and maybe harder to pinpoint exactly which PR caused the issue. So, it is again nice to do some testing upfront and not wait for release being cut.

Personally, I feel more confident with the work when it is tested. Code reviews catch some bugs but testing the change is always great when possible.

**Comment by b-slim** [comment]
Assuming this pr had +1 from @himanshug  I think it is a good pr so let's get it in

**Pull request merged**